src/lib.rs

```diff
@@ -198,13 +221,59 @@ fn num_bits<T>() -> usize {
     size_of::<T>() * 8
 }
 
-fn i_log2<T: PrimInt + Unsigned>(n: T) -> usize {
-    debug_assert!(n > T::zero());
-    let mut bits = num_bits::<T>() - 1;
-    while n & (T::one() << bits) == T::zero() {
-        bits -= 1;
+/// Convert (possibly signed) `x` of type `T` into an unsigned `StorageT`.
+fn abs<T, StorageT>(x: T) -> StorageT
+where
+    T: 'static + AsPrimitive<StorageT> + FromPrimitive + Ord + PrimInt + ToPrimitive,
+    StorageT: AsPrimitive<T> + PrimInt + Unsigned,
+{
+    debug_assert!(size_of::<StorageT>() >= size_of::<T>());
```

Comment:
This is where it starts getting tricky. The context is this:

```rust
    debug_assert!(size_of::<StorageT>() >= size_of::<T>());
    if x >= T::zero() {
        x.as_()
    } else if x == T::min_value() {
        T::max_value().as_() + StorageT::one()
    } else {
        (T::zero() - x).as_()
    }
```

The `debug_assert` is basically saying "any type that's X bits wide can represent another type of Y bits wide provided that Y >= X". That's clearly OK. The tricky bit, as you say, is why is `T::max_value().as_() + StorageT::one()` safe? It's because of the preceding `if x >= T::zero()`. If we have an unsigned type then this clause will always be true, so the `+ StorageT::one()` clause can never fire. In other words, the first `if` clause is for unsigned and signed types, but the second and third clauses are only for signed types.